bin/benchmarks/scanner.rb

```diff
@@ -0,0 +1,270 @@
+  test['real (significance)'] = significance(previous[label]['real'], test['real']) if previous && previous[label]
+  test['total (avg)'] = test['total'].reduce(:+) / test['total'].length
+  test['total (+/-)'] = previous && previous[label] &&
+    (test['total (avg)'] - previous[label]['total (avg)']) / test['total (avg)'] * 100
+  test['total (significance)'] = significance(previous[label]['total'], test['total']) if previous && previous[label]
+
+  test['real (variance)'] = test['real'].reduce(0) { |acc, value|
+    acc + (test['real (avg)'] - value) ** 2
+  } / test['real'].length
+  test['total (variance)'] = test['total'].reduce(0) { |acc, value|
+    acc + (test['total (avg)'] - value) ** 2
+  } / test['total'].length
+
+  test['real (sd)'] = Math.sqrt(test['real (variance)'])
+  test['total (sd)'] = Math.sqrt(test['total (variance)'])
+end
+
+log_data.push({
+  'time' => now,
+  'results' => results,
+})
+File.open(log, 'w') { |f| f.write(log_data.to_yaml) }
+
+def print_table(rows)
+  rows.each do |row|
+    row.each.with_index do |cell, i|
+      width = rows.reduce(0) { |acc, row| row[i].length > acc ? row[i].length : acc }
+      if i.zero?
+        print align(cell, width)
+      else
+        print(' ' + align(cell, width))
+      end
+    end
+    puts
+  end
+end
+
+def align(str, width)
+  if str.respond_to?(:justify)
+    case str.justify
+    when :center
+      ('%*s%s%*s' % [
+        ((width - str.length) / 2.0).round,
+        '',
+        str,
+        ((width - str.length) / 2.0).round,
+        '',
+      ])[0...width]
+    when :left
+      '%-*s' % [width, str]
+    else
+      '%*s' % [width, str]
+    end
+  else
+    '%*s' % [width, str]
+  end
+end
+
+AnnotatedString = Struct.new(:length, :to_s, :justify)
+def center(str)
+  AnnotatedString.new(str.length, str, :center)
+end
+
+def float(x)
+  '%.5f' % x
+end
+
+def parens(x)
+  "(#{x})"
+end
+
+def trim(str)
+  str.sub(/0+\z/, '')
+end
+
+def maybe(value, default = '')
+  if value
+    yield value
+  else
+    default
+  end
+end
+
+puts "\n\nSummary of cpu time and (wall-clock time):\n"
+
+headers = [
+  [
+    '',
+    center('best'),
+    center('avg'),
+    center('sd'),
+    center('+/-'),
+    center('p'),
+    center('(best)'),
+    center('(avg)'),
+    center('(sd)'),
+    center('+/-'),
+    center('p'),
+  ]
+]
+rows = headers + results.map do |(label, data)|
+  [
+    label,
+    float(data['total (best)']),
+    float(data['total (avg)']),
+    float(data['total (sd)']),
+    maybe(data['total (+/-)'], center('?')) { |value| '[%+0.1f%%]' % value },
+    maybe(data['total (significance)']) { |value| value > 0 ? trim(float(value)) : '' },
+    parens(float(data['real (best)'])),
+    parens(float(data['real (avg)'])),
+    parens(float(data['real (sd)'])),
+    maybe(data['real (+/-)'], center('?')) { |value| '[%+0.1f%%]' % value },
+    maybe(data['real (significance)']) { |value| value > 0 ? trim(float(value)) : '' },
+  ]
+end
+print_table(rows)
```

Comment:
Thanks for adding benchmarks! As this looks like mostly copy-pasta of the existing benchmarks it would be nice to pull the common bits out into a utility module so that they can be re-used without the duplication.